src/constraints.jl

```diff
@@ -182,31 +182,29 @@ end
 """
     CSRMatrix{T}
 
-Matrix given in compressed sparse row (CSR) format.
-
-`CSRMatrix` is analgous to the structure in Julia's `SparseMatrixCSC` but with
-the rows and columns flipped. It contains three vectors:
- - `row_pointers` is a vector pointing to the start of each row in
-    `columns` and `coefficients`;
- - `columns` is a vector of column numbers; and
- - `coefficients` is a vector of corresponding nonzero values.
+Type alias for Adjoint{T,SparseMatrixCSC{T,Int}} where T, whichc allows for the
+    efficient storage of CSR sparse matrix
+
+Please use the following interface to access this data incase model is changed in future
+ - `row_pointers(mat)` returns that sparse row_pointer vector see description
+    which has a length of rows+1
+ - `colvals` this is the column version of rowvals(SparseMatrixCSC)
+ - `coefficients` this is the equivilent to nonzeros(SparseMatrixCSC)
+"""
 
-The length of `row_pointers` is the number of rows in the matrix.
+const CSRMatrix{T} = Adjoint{T,SparseMatrixCSC{T,Int}} where T
 
-This struct is not a subtype of `AbstractSparseMatrix` as it is intended to be a
-collection of the three vectors as they are required by solvers such as Gurobi.
-It is not intended to be used for general computation.
-"""
-struct CSRMatrix{T}
-    row_pointers::Vector{Int}
-    columns::Vector{Int}
-    coefficients::Vector{T}
-    function CSRMatrix{T}(row_pointers, columns, coefficients) where T
-        @assert length(columns) == length(coefficients)
-        new(row_pointers, columns, coefficients)
-    end
+function CSRMatrix{T}(row_pointers, columns, coefficients) where T
+    @assert length(columns) == length(coefficients)
+    @assert length(columns) + 1 == row_pointers[end]
+    SparseMatrixCSC{T,Int}(maximum(columns), length(row_pointers)-1, row_pointers, columns, coefficients)'
 end
 
+#   Move access to an interface so that if we want to change type definition we can.
+colvals(mat::CSRMatrix) = rowvals(mat')
+row_pointers(mat::CSRMatrix) = mat'.colptr  # This is the only way to get at colptrs without recreating it
+row_nonzeros(mat::CSRMatrix) = nonzeros(mat')
```

Comment:
What not store `mat'` in the first place?

Reply:
`mat'`  on a type that is already wrapped in an adjoint is nearly free it takes like 1ns.   So by storing the data as Adjoint{SparseMatrixCSC} we are keeping the data in a scemantically correct way.   Furthermore, it is a good signature to dispatch on.